src/conn/conn_compact.c

```diff
@@ -18,13 +18,198 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
     return (FLD_ISSET(S2C(session)->server_flags, WT_CONN_SERVER_COMPACT));
 }
 
+/*
+ * __get_compact_stat --
+ *     Get the statistics for the given uri.
+ */
+static WT_BACKGROUND_COMPACT_STAT *
+__get_compact_stat(WT_SESSION_IMPL *session, const char *uri)
+{
+    WT_BACKGROUND_COMPACT_STAT *dsrc_stat;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t bucket, hash;
+
+    conn = S2C(session);
+
+    hash = __wt_hash_city64(uri, strlen(uri));
+    bucket = hash & (conn->hash_size - 1);
+
+    /* Find the uri in the files compacted list. */
+    TAILQ_FOREACH (dsrc_stat, &conn->background_compact.compacthash[bucket], hashq)
+        if (strcmp(uri, dsrc_stat->uri) == 0)
+            return (dsrc_stat);
+
+    return (NULL);
+}
+
+/*
+ * __should_compact --
+ *     Check whether we should proceed with calling compaction on the given file.
+ */
+static bool
+__should_compact(WT_SESSION_IMPL *session, const char *uri)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t cur_time;
+
+    conn = S2C(session);
+
+    /* The history store file should not be compacted. */
+    if (WT_STREQ(uri, WT_HS_URI))
+        return (false);
+
+    /* If we haven't seen this file before we should try and compact it. */
+    compact_stat = __get_compact_stat(session, uri);
+    if (compact_stat == NULL)
+        return (true);
+
+    /* If we have been unsuccessful recently skip this file for some time. */
+    cur_time = __wt_clock(session);
+    if (!compact_stat->prev_compact_success &&
+      WT_CLOCKDIFF_SEC(cur_time, compact_stat->prev_compact_time) < 60) {
```

Comment:
This 60 seconds is interesting - generally for magic numbers I prefer to make them a define.

I was thinking it would be more based on the amount of time spent during the compact on the particular file, not the amount of time between visits by the background compaction server.

Reply:
Yeah we can define this. We can also make it based on time spent during compaction, but we'll still need a method of deciding how long or how many times we should skip this file for.